### DISCUSSION TITLE
CORS error on offline invocation

### DISCUSSION THREAD

**Discussion by tjfoth** [BIPES/BIPES#166] [General]
Does anyone know how to get around this error when running in offline mode?

From the Developer's Javascript console: 

"Access to XMLHttpRequest at 'file:///Users/mine/Downloads/bipes_offline/ui/manifest.json' from origin 'null' has been blocked by CORS policy: Cross origin requests are only supported for protocol schemes: http, data, isolated-app, chrome-extension, chrome-untrusted, https, edge.manifest.json:1"

I get this in Microsoft Edge or Google Chrome.

I would very much like to use BIPES offline.

Thanks in advance for an answer!

**Comment by tjfoth** [comment]
Please disregard: I installed a local HTTP server.

**Comment by jetannenbaum** [comment]
Hi Tom,I’ve run BIPES offline without installing a local server.  You need to load the files as Java script files.  I can send you the code later this week if you would like to see the solution.   I don’t have access at the moment JimSent from Jet’s iPhoneOn Jan 2, 2023, at 5:35 PM, Tom Foth ***@***.***> wrote:﻿
Please disregard: I installed a local HTTP server.

**Reply by tjfoth** [reply]
I’d appreciate that!

**Comment by jetannenbaum** [comment]
Hi Tom,

  I'm back at home and created a branch called offline.  I briefly tested
the branch and it seems to run without issue on my local machine without a
webserver behind it.  Please note that I only have RP2040 based devices, so
the libraries I needed were only converted and tested on the pico, pico w
and maker pi boards.  If you want to use a different set of libraries, then
you will have to convert them to js files, include them in the
index_offline.html file and in the pylibsBlobs directory.  Also, modify the
code.js file to include the reference to the js blob files.

   Please let me know if you have any issues or questions,

    Jim

**Comment by tjfoth** [comment]
Jim,

Thanks so much and I'll take a look this later this evening.

Like you, I'm focused on RP2040.

Since we're talking about RP2040: Have you done anything with Timers?  The RP2040 micro python code wants a -1 for the Timer id, because the timers are all virtual (software). The BIPES UI won't let me put in -1.

What I've been doing is either:
1. Using the Timer blocks and then manually fixing the code, or
2. Setting up the timers with the Python block.

Tom

**Comment by jetannenbaum** [comment]
Hi Tom,

   I haven't played with timers in BIPES, but it is straight forward to
extend the BIPES code to allow a -1 id for the timer.  Is it simply a
matter of allowing -1, or do you always want a -1 as the timer id?
Assuming you want this to be RP2040 specific, and you are running your own
version either is fine.

   Jim

**Comment by jetannenbaum** [comment]
Hi Tom,

   I looked a little deeper into this.  There already is a block definition
for the pico_timer.  Should be as simple as updating the timer block in the
rpi_pico.xml, rpi_pico_w.xml, and makerpi.xml files in the toolbox folder.
Then rebuild the index_offline.html file.  You will probably want to clean
up the stop_timer block as well, since the RP2040 doesn't require the timer
id to stop the virtual timer.

   Jim

**Reply by jetannenbaum** [reply]
Hi Tom,

   One more update.  You probably want to create a pico_stop_timer block that contains the import for  the timer and returns code tim.deinit() to shut down the timers you create.  You will need to call the pico_stop_timer block in your rpi_pico.xml, rpi_prico_w.xml, and makerpi.xml files, replacing the stop_timer block.  Then create the block in the block_definitions.js file and add the code in the generator_stubs.js file (both found in the core folder).  Once again, rebuild the index_offline folder when done.

**Comment by tjfoth** [comment]
I’m going to look into this during this weekend.  I’ll put it first on the
fork you created.

On Sat, Jan 7, 2023 at 11:23 AM Jim Tannenbaum ***@***.***>
wrote:

> Hi Tom,
>
> One more update. You probably want to create a pico_stop_timer block that
> contains the import for the timer and returns code tim.deinit() to shut
> down the timers you create. You will need to call the pico_stop_timer block
> in your rpi_pico.xml, rpi_prico_w.xml, and makerpi.xml files, replacing the
> stop_timer block. Then create the block in the block_definitions.js file
> and add the code in the generator_stubs.js file (both found in the core
> folder). Once again, rebuild the index_offline folder when done.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/BIPES/BIPES/discussions/166#discussioncomment-4620642>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOIIO2RRJGFMML7KA3G6VTWRGJ6HANCNFSM6AAAAAATONEIVI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
—-
Tom Foth

***@***.***
